design/design-doc-composition.md

```diff
@@ -0,0 +1,1079 @@
+      transforms:
+      - type: math
+        math:
+          multiply: 1024
+    - fromFieldPath: "spec.region"
+      toFieldPath: "spec.forProvider.location"
+      transforms:
+      - type: map
+        map:
+          us-west: "West US"
+          us-east: "East US"
+    # Specifies the (potentially sensitive) connection details that this 'to'
+    # resource should expose to the 'from' resource. Names are unique across all
+    # 'to' resources within this composition. Ignored by application resources.
+    connectionDetails:
+    - name: username
+      fromConnectionSecretKey: username
+    - name: password
+      fromConnectionSecretKey: password
+    - name: endpoint
+      fromConnectionSecretKey: endpoint
+  - base:
+      apiVersion: database.azure.crossplane.io/v1alpha3
+      kind: MySQLServerVirtualNetworkRule
+      spec:
+        name: my-cool-vnet-rule
+        serverNameSelector:
+          matchComposite: true
+        resourceGroupNameSelector:
+          matchComposite: true
+        properties:
+          virtualNetworkSubnetIdRef:
+            name: sample-subnet
+        reclaimPolicy: Delete
+        providerRef:
+          name: azure-provider
+```
+
+This document proposes compositions be immutable. This property would allow
+updates to a 'from' resource to be propagated to the various 'to' resources but
+keep the set of 'to' resources stable over the lifetime of the 'from' resource,
+allowing controllers to avoid complex update and garbage collection logic.
+
+### ApplicationDefinition
+
+A `kind: ApplicationDefinition` defines a new kind of custom resource that
+represents an application - for example a `kind: Wordpress` custom resource. An
+application resource is namespaced, and may only compose resources within its
+namespace. Applications are typically defined by an _application developer_ -
+the developer of Wordpress would author the `kind: ApplicationDefinition` that
+defines the schema of a `kind: Wordpress` resource. Application resources are
+typically authored by an _application operator_.
+
+Here's an example that defines a new application resource of `kind: Wordpress`:
+
+```yaml
+apiVersion: crossplane.io/v1alpha1
+kind: ApplicationDefinition
+metadata:
+  # ApplicationDefinition names are subject to the constraints of Kubernetes
+  # CustomResourceDefinition names. They must be of the form <plural>.<group>.
+  name: wordpresses.apps.example.org
+spec:
+  # Defines the structural schema and GroupVersionKind of this application. Only
+  # a single API version of the application may exist. Additional fields will be
+  # injected to support composition machinery.
+  crdSpecTemplate:
+    group: apps.example.org
+    version: v1alpha1
+    names:
+      kind: Wordpress
+      listKind: WordpressList
+      plural: wordpresses
+      singular: wordpress
+    validation:
+      openAPIV3Schema:
+        properties:
+          administratorLogin:
+            type: string
+          storageSize:
```

Comment:
I am not sure why the application model has storage-related field as top-level properties. I assume that they have to be matched to some infrastructure field eventually.

Reply:
Keep in mind that this is just an example of what an application developer might choose to include in their schema. Some could decide that the amount of storage a DB has is not relevant to an application operator, but I suspect many would pick this field as something the app operator may care about.